lib/semantics/tools.cc

```diff
@@ -508,6 +509,139 @@ std::optional<parser::MessageFixedText> WhyNotModifiable(
+    const parser::SyncMemoryStmt &) {
+  return "SYNC MEMORY is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::SyncTeamStmt &) {
+  return "SYNC TEAM is not allowed in this context"_err_en_US;
+}
+
+bool IsCoarray(const parser::AllocateObject &allocateObject) {
+  const parser::Name &name{GetLastName(allocateObject)};
+  if (name.symbol && IsCoarray(*name.symbol)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::AllocateStmt &allocateStmt) {
+  const auto &allocationList{
+      std::get<std::list<parser::Allocation>>(allocateStmt.t)};
+  for (const auto &allocation : allocationList) {
+    const auto &allocateObject{std::get<parser::AllocateObject>(allocation.t)};
+    if (IsCoarray(allocateObject)) {
+      return "ALLOCATE of a coarray is not allowed in this context"_err_en_US;
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::DeallocateStmt &deallocateStmt) {
+  const auto &allocateObjectList{
+      std::get<std::list<parser::AllocateObject>>(deallocateStmt.t)};
+  for (const auto &allocateObject : allocateObjectList) {
+    if (IsCoarray(allocateObject)) {
+      return "DEALLOCATE of a coarray is not allowed in this "
+             "context"_err_en_US;
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::ChangeTeamStmt &) {
+  return "CHANGE TEAM is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::CriticalStmt &) {
+  return "CRITICAL is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::EventPostStmt &) {
+  return "EVENT POST is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::EventWaitStmt &) {
+  return "EVENT WAIT is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::FormTeamStmt &) {
+  return "FORM TEAM is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::LockStmt &) {
+  return "LOCK is not allowed in this context"_err_en_US;
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::UnlockStmt &) {
+  return "UNLOCK is not allowed in this context"_err_en_US;
+}
+
+bool HasCoarray(const parser::Expr &expression) {
+  if (const auto *expr{GetExpr(expression)}) {
+    for (const Symbol *symbol : evaluate::CollectSymbols(*expr)) {
+      if (const Symbol * root{GetAssociationRoot(DEREF(symbol))}) {
+        if (IsCoarray(*root)) {
+          return true;
+        }
+      }
```

Comment:
Can this be written as
```
const Symbol* root{GetAssociationRoot(DEREF(symbol))};
return root && IsCoarray(*root);
```
to reduce nesting?

Reply:
I don't think that would work.  This test is inside a loop that covers all of the symbols from the expression.  Your suggested change would cause the loop to exit at the first symbol.

Reply:
Good point, I think my internal optimizer was a bit overzealous here!
We could use some C++17 niceness to reduce nesting here with the same behaviour:
```
if (const Symbol * root{GetAssociationRoot(DEREF(symbol))}; root && IsCoarray(*root)) {
  return true;
}
```
but it's not really necessary

Reply:
Yep.  For me, the nesting is more readable, even though it uses more lines.